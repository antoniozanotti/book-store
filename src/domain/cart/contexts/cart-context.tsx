"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
