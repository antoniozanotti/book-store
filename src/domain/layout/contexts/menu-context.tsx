"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
