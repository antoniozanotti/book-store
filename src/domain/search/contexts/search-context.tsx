"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
