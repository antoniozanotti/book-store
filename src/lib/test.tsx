import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
