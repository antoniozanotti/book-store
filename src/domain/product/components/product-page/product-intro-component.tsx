"use client";

import { UiTitleComponent } from "@/domain/ui/components/ui-title-component";
import { useProductPageContext } from "../../contexts/product-page-context";
import { AuthorsWithLinkComponent } from "./authors-with-link-component";
import { ProductWithBookAndCategoriesPrisma } from "../../types/product-prisma";
import { UiFigureComponent } from "@/ui-figure-component";

export function ProductIntroComponent({
  ...props
}: React.ComponentProps<"div">) {
  const product: ProductWithBookAndCategoriesPrisma =
    useProductPageContext().state.product;

  return (
    <div {...props}>
      <UiTitleComponent size="large">{product.name}</UiTitleComponent>
      {product.book && (
        <AuthorsWithLinkComponent authors={product.book.authors} />
      )}
      <UiFigureComponent
        src={`product/${product.page.slug}`}
        width={374}
        height={497}
        alt={product.name}
        className="w-full sm:w-[374px] h-[497px] relative overflow-hidden mt-4"
        innerImgStyle="object-cover w-full h-full"
      />
      <UiTitleComponent level="h2" size="medium" className="mt-10">
        Synopses
      </UiTitleComponent>
      <p>{product.description}</p>
    </div>
  );
}
