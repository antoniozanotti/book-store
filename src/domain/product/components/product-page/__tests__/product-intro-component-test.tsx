import { render, screen } from "@/lib/test";
import { ProductIntroComponent } from "../product-intro-component";
import { ProductPageProvider } from "@/domain/product/contexts/product-page-context";
import { productWithBookAndCategoriesMock } from "@/domain/product/__tests__/product-with-book-and-categories-mock";

describe("ProductIntroComponent", () => {
  it("should render", async () => {
    const { container } = render(
      <ProductPageProvider product={productWithBookAndCategoriesMock}>
        <ProductIntroComponent />
      </ProductPageProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
