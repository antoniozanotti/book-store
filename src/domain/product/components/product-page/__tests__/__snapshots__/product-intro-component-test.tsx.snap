// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductIntroComponent should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <div>
      <h1
        class="font-primary text-pretty text-4xl sm:text-5xl md:text-6xl lg:text-7xl"
      >
        The Four Agreements: A Practical Guide to Personal Freedom
      </h1>
      <address>
        by 
        <a
          class="underline"
          href="/author/don-miguel-ruiz"
          rel="author"
          title="Don Miguel Ruiz"
        >
          Don Miguel Ruiz
        </a>
      </address>
      <figure
        class="bg-green-300 w-full sm:w-[374px] h-[497px] relative overflow-hidden mt-4"
      >
        <img
          alt="The Four Agreements: A Practical Guide to Personal Freedom"
          class="transition-all object-cover w-full h-full"
          data-nimg="1"
          decoding="async"
          height="497"
          loading="lazy"
          sizes="374px"
          src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5102,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0"
          srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_21,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_42,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_63,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_85,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_127,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_170,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_340,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_510,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_850,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_996,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1100,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1435,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1594,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2551,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2721,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5102,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 3840w"
          style="color: transparent;"
          width="374"
        />
      </figure>
      <h2
        class="font-primary text-pretty text-3xl mt-10"
      >
        Synopses
      </h2>
      <p>
        Canto eum versus somniculosus mollitia conforto celebrer iste. Surculus facere vix acceptus. Cohaero demo sufficio venio autus animi ex adversus quam.
      </p>
    </div>
  </div>
</div>
`;
